File: amazon/692_top_k_frequent_word.cpp

class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        // Frequency map to count occurrences of each word
        unordered_map<string, int> freq;
        
        // Count frequency of each word
        for (vector<string>::const_iterator it = words.begin(); it != words.end(); ++it) {
            const string& word = *it;
            freq[word]++;
        }
        
        // Custom comparator for the priority queue
        struct Comparator {
            bool operator()(const pair<int, string>& a, const pair<int, string>& b) {
                if (a.first == b.first) {
                    return a.second > b.second; // Sort lexicographically if frequencies are equal
                }
                return a.first < b.first; // Sort by frequency in descending order
            }
        };
        
        // Declare the priority queue with the custom comparator
        priority_queue<pair<int, string>, vector<pair<int, string>>, Comparator> pq;
        
        // Push elements into the priority queue
        for (unordered_map<string, int>::const_iterator it = freq.begin(); it != freq.end(); ++it) {
            pq.emplace(it->second, it->first);
        }
        
        // Extract the top k elements
        vector<string> result;
        while (k-- > 0 && !pq.empty()) {
            result.push_back(pq.top().second);
            pq.pop();
        }
        
        return result;
    }
};

File: amazon/692_top_k_frequent_word.cs
public class Solution {
    public IList<string> TopKFrequent(string[] words, int k) {
        Dictionary<string, int> freq = new Dictionary<string, int>();
        foreach(var s in words){
            if(freq.ContainsKey(s)) freq[s]++;
            else freq.Add(s, 1);
        }
        var resList = freq.ToList();
        resList.Sort((p1, p2)=>{
            if(p2.Value - p1.Value == 0) return String.Compare(p1.Key, p2.Key);
            return p2.Value - p1.Value;
        });
        List<string> ans = new List<string>();
        int i = 0;
        foreach(var n in resList){
            ans.Add(n.Key);
            if(++i == k) return ans;
        }
        return null;
    }
}